### PR TITLE
feat: makes the library esm-compatible

### LINKED ISSUE #1267: Can't resolve '@mswjs/interceptors/lib/interceptors/ClientRequest'
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Browsers

NODE 16/18

### Reproduction repository

https://github.com/coinbase/rest-hooks/tree/a3963b2618981992bab115a8e8543abdb9cdc655/website

Gonna revert MSW upgrade to make this work so master will no longer have this problem.

### Reproduction steps

1. clone repo above
2. git checkout a3963b2618981992bab115a8e8543abdb9cdc655
3. navigate to website folder
4. yarn install
5. yarn start
6. observe error in cli as well as opened browser tab

### Current behavior

```bash
ERROR in ./node_modules/msw/lib/node/index.mjs 45:0-94

Module not found: Error: Can't resolve '@mswjs/interceptors/lib/interceptors/ClientRequest' in '/home/ntucker/src/rest-hooks/website/node_modules/msw/lib/node'
Did you mean 'index.js'?
BREAKING CHANGE: The request '@mswjs/interceptors/lib/interceptors/ClientRequest' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
```

https://github.com/mswjs/msw/pull/1247 breaks in node versions that support ESM

### Expected behavior

To fix this - the ESM version of MSW must be explicit about file extensions. Otherwise it is by definition incorrectly built. If this cannot be done you should not ship a custom ESM version as this is incorrect. The only reason it would work is if it is used by a system that doesn't enforce fully (webpack). However this will break in fully-spec compliant things like node.

More simply:

`import '@mswjs/interceptors/lib/interceptors/ClientRequest'` => `import '@mswjs/interceptors/lib/interceptors/ClientRequest.mjs'`

### PR DESCRIPTION
**Pull request by ivanhofer**
fixes #1267

I have investigated #1267 a bit. There are a few road blocks to make `msw/node` work in a `.mjs` file:
 - [x] reference `lib/index.js` instead of `'lib'` in a few files (easy to achive)
 - [x] ~`tsup` does not support import assertions (needed here https://github.com/mswjs/msw/blob/main/src/context/status.ts#L1) https://github.com/egoist/tsup/issues/714~
    workaround: https://github.com/mswjs/msw/pull/1399#discussion_r968360168
 - [x] ~dynamic import of `node-fetch` (https://github.com/mswjs/msw/blob/main/src/context/fetch.ts#L6) is not supported in esm context. https://github.com/evanw/esbuild/issues/1921~
    workaround: https://github.com/mswjs/msw/pull/1399#issuecomment-1243681143

### CODE REVIEW

File: src/context/status.ts
Comment (ivanhofer) [inline src/context/status.ts]: this will currently be removed on build

Comment (mattcosta7) [reply src/context/status.ts]: we could also configure `tsup` to inline this library by moving it to `devDependencies` instead of listing in dependencies

File: src/node/setupServer.ts
Comment (mattcosta7) [inline src/node/setupServer.ts:2]: we should really just update interceptors to expose an esm build.  

I started on that over the weekend, and have some thoughts on how we could do that, but I think we'll need to probably remove `debug` as a dependency there - since it relies on `tty` and `os`, and i'm having issues getting those properly bundled to a browser compat version, that multiple build tools like.

File: src/handlers/GraphQLHandler.test.ts
Comment (mattcosta7) [inline src/handlers/GraphQLHandler.test.ts]: When I get some time, exposing an esm build of headers-polyfill should be pretty quick, and might be better than importing from lib directly here

**Comment by mattcosta7** [comment]
> dynamic import of node-fetch

I think we could get around this with something like

```
useFetch = isNodeProcess() ? (input, init) => import('node-fetch').then(({default: nodeFetch}) => fetch(input, init)) : window.fetch
```

I'm curious if we could also avoid using node-fetch directly in this library and just stipulate that `globalThis.fetch` be defined already (via a user polyfill) instead?

File: src/handlers/GraphQLHandler.test.ts
Comment (ivanhofer) [reply src/handlers/GraphQLHandler.test.ts]: Yes this would be the cleanest solution. Until then this would be an easy workaround

**Comment by ivanhofer** [comment]
With this branch it is now poosible to run the following `esm` file:

_test.mjs_
```js
import { setupServer } from 'msw/node'

setupServer()
```

with the command

`node ./test.mjs`

File: src/handlers/GraphQLHandler.test.ts
Comment (mattcosta7) [reply src/handlers/GraphQLHandler.test.ts]: just opened https://github.com/mswjs/headers-polyfill/pull/38

Comment (kettanaito) [reply src/handlers/GraphQLHandler.test.ts]: `headers-polyfill@3.1` ships with ESM support now! Thank you, @mattcosta7! 🎉 

I've updated `@mswjs/interceptors` to use it, now updating MSW as well.

File: src/node/setupServer.ts
Comment (kettanaito) [reply src/node/setupServer.ts:2]: We need to look into some `debug` recipes perhaps. It has browser support, so it should run in the browser. Perhaps we can do some import map for this dependency to ensure that in the browser it always imports the browser dist.

Comment (mattcosta7) [reply src/node/setupServer.ts:2]: Yeah, it mostly relates to the fact that `debug` doesn't have an esm build, so loading in a browser directly from `@web/dev-server` seems problematic.  

Maybe that just means users in those environments have to do some extra setup, and that's enough.  Will keep playing with configurations there, because it's also possible we don't need to be as explicit there.

I was testing using tsup to bundle, however using tsc to generate cjs and esm transpilations might help avoid issues related to that (since it won't try to handle debug directly)

File: src/handlers/GraphQLHandler.test.ts
Comment (ivanhofer) [reply src/handlers/GraphQLHandler.test.ts]: I just did a rebase

**Comment by filipw01** [comment]
Nice, work. I was about to open an issue, but it looks like it should be fixed once this PR is merged 😄

**Comment by ivanhofer** [comment]
I think this PR is ready to be merged. Or is something still missing?

**Comment by mataspetrikas** [comment]
@kettanaito @ivanhofer is this coming into the next release?

**Review by kettanaito — APPROVED** [review]
Finally got to this improvement. Looks superb 🎉 Updated against the latest main, waiting for the CI before marking it as a release candidate. Expect to land in the next minor release. 

Thank you for improving MSW, @ivanhofer! If you ever feel like exercising your skill again, we could also migrate the [interceptors](https://github.com/mswjs/interceptors) library to ESM.

**Comment by kettanaito** [comment]
The only thing I'd add is an automated test for the distributed ESM format. Let me see if I can dig out the previous test I've written...

**Comment by kettanaito** [comment]
## Released: v0.48.0 🎉

This has been released in v0.48.0!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v0.48.0)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/0.48.0)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Pull request merged**
